profiling/src/php_ffi.c

```diff
@@ -75,9 +75,53 @@ static post_startup_cb_result ddog_php_prof_post_startup_cb(void) {
 static bool _ignore_run_time_cache = false;
 #endif
 
-void datadog_php_profiling_startup(zend_extension *extension) {
+#if PHP_VERSION_ID >= 70400 && PHP_VERSION_ID < 80100
+/* The purpose here is to set the opline, because these versions are missing a
+ * SAVE_OPLINE() and it causes a crash for the allocation profiler, see:
+ * https://github.com/php/php-src/commit/26c7c82d32dad841dd151ebc6a31b8ea6f93f94a
+ * The handler doesn't actually need to do anything, because just by having a
+ * handler the engine will save the opline before calling the user handler:
+ * https://heap.space/xref/PHP-7.4/Zend/zend_vm_execute.h?r=0b7dffb4#2650
+ */
+static zend_result ddog_php_prof_opcode_dispatch(zend_execute_data *execute_data) {
+    return ZEND_USER_OPCODE_DISPATCH;
+}
+#endif
 
-#if CFG_RUN_TIME_CACHE // defined by build.rs
+static void ddog_php_prof_install_opcode_handlers(uint32_t php_version_id) {
+#if PHP_VERSION_ID < 70400 || PHP_VERSION_ID >= 80100
+    (void)php_version_id;
+#else
+
+    /* Only need to install handlers if there isn't one, see docs on
+     * `ddog_php_prof_opcode_dispatch`.
+     */
+
+    /* Issue is fixed in 8.0.26:
+     * https://github.com/php/php-src/commit/26c7c82d32dad841dd151ebc6a31b8ea6f93f94a
+     * But the tracer installed a handler for ZEND_GENERATOR_CREATE on PHP 7,
+     * so nearly no users will triggered this one.
+     */
+    if (php_version_id < 80026 && zend_get_user_opcode_handler(ZEND_GENERATOR_CREATE) == NULL) {
+        user_opcode_handler_t handler = (user_opcode_handler_t)ddog_php_prof_opcode_dispatch;
+        zend_set_user_opcode_handler(ZEND_GENERATOR_CREATE, handler);
+    }
+
+    /* Issue is fixed in 8.0.12:
+     * https://github.com/php/php-src/commit/ec54ffad1e3b15fedfd07f7d29d97ec3e8d1c45a
+     * The tracer didn't save us here. Fortunately, throwing destru
```

Comment:
todo: fix this comment lol

Reply:
Oh I remember. It is not only about fixing the comment, but try and see if the fix is enough, as there is a `zend_array_dup()` call few lines before the `SAVE_OPLINE()` happens. I'll check if I can trigger this, I highly assume it will crash as well.

Reply:
Yes, we need to adjust the version range too. If it's still on PHP 8.3, then we need to install the opcode handler on more versions too.

Reply:
Jup, we need to install up to latest, can you check [ad46787](https://github.com/DataDog/dd-trace-php/pull/2352/commits/ad467876e9bad7f9cb9e826db595619ce5ee911d)

Reply:
Looks good. I add another commit that cleans things up and adds a pointless test for `ZEND_BIND_STATIC` with a todo.